### DISCUSSION TITLE
Question: Best way model a divers/pluriform sensor web

### DISCUSSION THREAD

**Discussion by rduivenvoorde** [FraunhoferIOSB/FROST-Server#537] [Q&A]
Hi, thinking about using STA to keep/query a database which keeps all kind of measurements (in all kind of substances) from all kind of different research institutes (substance here are the 'stuff that you are measuring in' being food- or animal-types):

In STA terms (?) multiple (tens?) of different ObservedProperties,
- physical quantity A in substance A
- physical quantity A in substance B
- physical quantity B in substance A
- physical quantity C in substance A
etc

As http://www.qudt.org/qudt/owl/1.0.0/quantity/Instances.html seems not to be specific enough, we thought about creating a uri-schema ourselves:
http://example.com/def/quantityA/substanceA
http://example.com/def/quantityA/substanceB
etc..
Let's define we will have around a hundred quantity/substance types.

Question 1: would this be a feasible model to use, or has somebody better experience/idea's?

Question 2: would it be possible to query only for Observations from a given quantity/substance type combi (one ObservedProperty), preferably in a given area and time frame ... ?

Thanks in advance for any input.

**Comment by mjacoby** [comment]
As your questions rather concern the STA standard itself than the FROST Server implementation you might consider re-posting it in the STA project itself to gain higher visibility https://github.com/opengeospatial/sensorthings

Nevertheless, I'll try to answer your questions as good as I can. 

**Question 1**
The problem of modelling Properties and Feature of Interest (FoI) of an observation has been around for a long time even before STA. It seems that there is not one perfect solution but multiple viable ways of doing this.
For example in the SSN/SOSA ontology they advocate to create an instance of ssn:Property for each combination of FoI and actual/generic property (e.g. temperature) and use this as property of the observation. They also do this for lat/long based FoIs which would create properties with a name like this 

> Air?lat=45.75&long=4.85#temperature

Although it is possible, I'm not sure if this is a good idea to use the same approach with STA as it would create many ObservedProperty entities that are not very likely to be re-used among senors.  I would suggest to use your approach creating one ObservedProperty per combination of physical quantity and substance. I would recommend using the properties property of each ObservedProperty entity with a custom notation to store which physical quantity and substance it is referring to, e.g. 
```
{
	"name": "Temperature in Air",
	"definition": "...",
	"properties": {
		"type": "composed",
		"physicalQuantity": "temperature",
		"substance": "air"
	}
}
```


**Question 2**
Yes, that is possible. Given the above way to store your data a possible query could look like this

```
http://example.org/FROST-Server/v1.0/Observations?
     $filter=Datastream/ObservedProperty/properties/physicalQuantity eq 'temperature' and
     Datastream/ObservedProperty/properties/substance eq 'air' and 
     st_within(FeatureOfInterest/feature, geography'POLYGON ((30 10, 10 20, 20 40, 40 40, 30 10))') and 
     phenomenonTime gt 2018-01-01T00:00:00Z and 
     phenomenonTime lt 2018-01-02T00:00:00Z
```

There are other spatial functions available in STA, see the corresponding section of the standard document at http://docs.opengeospatial.org/is/15-078r6/15-078r6.html#56.

**Comment by rduivenvoorde** [comment]
Ah, thanks @mjacoby for the pointer and answers!

We were not aware of the fact we could also use properties this way. And as long as you keep the ObservedProperties(table) count low, it will still perform ok without the possibility to put an index on this 'properties\-table. 

(Performance is great anyway by the way!)

**Comment by mjacoby** [comment]
Be aware that with v1.0 of STA ObservedProperty does not have the properties property - that is (at the moment) something FROST-specific. However, it will be part of v1.1 of the standard which hopefully will be available later this year (https://github.com/opengeospatial/sensorthings/issues/49).

FYI, you can of course modify your database and add additional indices as you wish if you host your own server. This can significantly reduce query execution times in some cases. As this is specific to each deployment and depending on the type of queries used, FROST Server ships only with a few basic indices by default.

**Comment by hylkevds** [comment]
There are indeed many different ways to apply the STA data model. What do you use as Thing in your use-case? Some other options would be:

- Use the substance as the Thing. This means you don't need separate ObservedProperties for the same Property in difference substances.
- Use the FeatureOfInterest as the substance. This has the same advantage,

Either way, you should not run into performance issues. For BRGM we've converted the French surface water quality database to STA. It has 1874 Observed Properties, 18478 Things, 9.5 million Datastreams, 136 million Observations and 1.9 million FeaturesOfInterest, and performance is fine.

**Comment by KathiSchleidt** [comment]
And - there are definite advantages to keeping your observedProperty separate from your measurement matrix (what you're measuring the property in, Substance A), i.e. you can easily query for physical quantity X across Substances (and you keep your observedProperty managable). And, as Hylke has pointed out, we just proved to BRGM that "yes we can!" ;) (in this deployment, the measurement matrix is a property of the FeatureOfInterest, and the queries work)
I'd be careful about assigning the matrix to the Thing as

1. not semantically sound (semantically this is a property of the FoI)
2. constrains reuse of the Thing (I could see use cases where one Thing hosts multiple Sensors, which in turn analyse various samples (FoIs))

There is a further conundrum pertaining to observedProperty - how you define physical quantity X. This is actually also a complex concept, usually breaks down into something physical (i.e. a chemical, but could also be a species) and some measure type (i.e. concentration (usual for chemicals), or count (more suitable to species)). In some domains this is automatically subsumed (when one speaks of Ozone pertaining to Air, everybody knows it's Concentration of Ozone), in other must be made explicit (i.e. count of a specific species)

**Comment by rduivenvoorde** [comment]
@hylkevds @KathiSchleidt thanks for this input. One drawback I can see for us in this (use FOI) setup (we are very much location oriented) is that  the data provider then itself has to create the FOI"s (aka: get the location and put them in the FOI)?  While now: this is done by the Frost itself, so we get the FOI's for free. Though it is not clear yet to us, WHICH location Frost is putting in the FOI: the location at the moment the Observation is POSTed?

But indeed, FOI does not have so much information it itself, if you do not care too much about locations.

( I do have troubles changing the location from a thing though, but will make that another question :-) )

To me, the hard part of using STA is how to model the different use-cases into it, there are A LOT of possible solutions, and like with a database model, the more normalisation in your model, the more flexible it becomes, but also harder to understand. Maybe some practical use-cases (like the BRGM case?) written down would help to decide for the model for newcomers. Practical queries or setups are hard to find/google... (I see that the issues of sensorthings repo contains 'bestpractices' labels, eg about moving Things: https://github.com/opengeospatial/sensorthings/issues/33

FYI, for me the cases are:
- more or less static sensors spread over Europe, measuring a couple of physical quantities (in air)
- driving car(s) which measure more or less the same quantities (OR go into the field to pick some grass or other substances)
- incidental/accidental measurements from governmental institutes (of which the actual place of the measurement is the FOI location)

(sorry for bringing this up hear, I agree with @mjacoby that I probably should do this on sensorthings level....)

**Comment by mjacoby** [comment]
The standard sas the following about generated FoI upon inserting an observation

> In this case, when a client creates an Observation entity, the client SHOULD omit the link to a FeatureOfInterest entity in the POST body message and SHOULD not create a related FeatureOfInterest entity with deep insert. And if the service detects that there is no link to a FeatureOfInterest entity in the POST body message that creates an Observation entity, the service SHALL either (1) create a FeatureOfInterest entity by using the location property from the Location of the Thing entity when there is no FeatureOfInterest whose location property is from the Location of the Thing entity or (2) link to the FeatureOfInterest whose location property is from the Location of the Thing entity.

http://docs.opengeospatial.org/is/15-078r6/15-078r6.html#60

In FROST it is implemented according to it 
https://github.com/FraunhoferIOSB/FROST-Server/blob/97ac5f9a7c4184dda58b209b0a392b10a80ab18f/FROST-Server.SQL/src/main/java/de/fraunhofer/iosb/ilt/frostserver/persistence/postgres/EntityFactories.java#L354-L439

So actual behaviour is to take the location of the corresponding thing at the time of creation/insert of a new observation. However, this might actually not be the only valid option. It could probably make sense to take the location of the thing at the time given by Observation.resultTime, e.g. in scenarios where observations are only inserted every now and then in batch style (but position of the Thing is updated in "real-time"). This might be worth talking about it I guess. Although it would/should be possible to explicitely set the FoI when inserting observations in such a batch style this seems to be unnecessary complex as it could easily be done automatically.
I guess this is also something to discuss at SensorThings level.

**Comment by hylkevds** [comment]
When an Observation is POSTed that has no FoI information in the POST, FROST will check the Location of the Thing of the Datastream the Observation is POSTed to:

1. If a FoI has already been generated for this Location, this FoI is used
2. If no FoI is generated yet for this Location, a new FoI is generated, and used.

**Comment by rduivenvoorde** [comment]
@hylkevds ok, clear. Thanks. 
Working with moving measurement vehicles (including portable off line probes in them), I was wondering how this relates to https://github.com/opengeospatial/sensorthings/issues/30 (in spec V1.1?) where it is then possible to change/add the HistoricalLocation of the Thing.

But what happens to the FOI then? Is the feature/location in it, determined based on the (latest) Location then?
Or is there some heuristic to determine the location of the Observation based on phenomenonTime or resultTime  and the HistoricalLocations?
Or is the order of creation of objects leading?

Or do I miss some point here?

**Comment by hylkevds** [comment]
When a FoI is generated it is always based on the current Location object of the Thing. There is no heuristic based on phenomenonTime. So when there is a backlog of both location and sensor data, as in https://github.com/opengeospatial/sensorthings/issues/30, the order in which the Observations and Locations are sent is important.

In certain cases it might be easier to not use the auto generated FoIs but manually generate FoIs instead.
